src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -1482,8 +1482,8 @@ bool LibraryCallKit::inline_vector_gather_scatter(bool is_scatter) {
     }
 
     // Check whether the predicated gather/scatter node is supported by architecture.
-    if (!arch_supports_vector(is_scatter ? Op_StoreVectorScatterMasked : Op_LoadVectorGatherMasked, num_elem, elem_bt,
-                              (VectorMaskUseType) (VecMaskUseLoad | VecMaskUsePred))) {
+    VectorMaskUseType mask = (is_scatter || !is_subword_type(elem_bt)) ? (VectorMaskUseType) (VecMaskUseLoad | VecMaskUsePred) : VecMaskUseLoad;
+    if (!arch_supports_vector(is_scatter ? Op_StoreVectorScatterMasked : Op_LoadVectorGatherMasked, num_elem, elem_bt, mask)) {
```

Comment:
sub-word gather do not emit any predicated instructions, thus only VectorMaskUseLoad is relevant in this context, however AVX512 and SVE does have a direct predicated gather instructions for 32/64 bit types.

Reply:
I see, thanks! `VecMaskUsePred` is added to check `match_rule_supported_vector_masked` for normal vector ops. That's because we may add an additional mask input for those vector ops. But `Load|StoreVectorScatterMasked` are different. They point to the masked operations no matter how they are implemented. So just `VecMaskUseLoad` is fine for all these two ops for me.

Reply:
I think its better to align masked sub-word gather implementation with non-sub-word once i.e. support intrinsfication only for  predicated targets. Respective backends may then choose to either emit a predicated loop like the one which this patch does OR directly emit a predicated instruction if target support it.

With this we may see some performance penalty for masked sub-word gathers on non-predicated targets since original java implementation will now become fallback code, but then same penalty exists for non-subword gathers today.

Reply:
> I think for the time being its better to align masked sub-word gather implementation with non-sub-word ones i.e. support intrinsfication only for predicated targets. Respective backends may then choose to either emit a predicated loop like the one which this patch does OR directly emit a predicated instruction if target support it. WDYT?

Agree!